Google.Api.Gax.Testing/FakeScheduler.cs

```diff
@@ -180,78 +195,166 @@ public void Run<T>(Func<Task<T>> taskProvider)
         public async Task<T> RunAsync<T>(Func<Task<T>> taskProvider)
         {
             var funcTask = Task.Run(taskProvider);
-            var simulatedTimeTask = StartLoopAsync(Clock.GetCurrentDateTimeUtc() + SimulatedTimeTimeout);
+            var simulatedTimeTask = StartLoopAsync(Clock.GetCurrentDateTimeUtc() + SimulatedTimeTimeout, true);
             var delayTask = Task.Delay(RealTimeTimeout);
-            var completedTask = await Task.WhenAny(funcTask, simulatedTimeTask, delayTask).ConfigureAwait(false);
+            await Task.WhenAny(funcTask, simulatedTimeTask, delayTask).ConfigureAwait(false);
+            // Allow the test code a little more time to execute, just in case we're waiting for slow logging etc.
+            await Task.WhenAny(funcTask, Task.Delay(PostLoopSettleTime)).ConfigureAwait(false);
             lock (_monitor)
             {
                 _stopped = true;
                 Monitor.PulseAll(_monitor);
             }
-            if (completedTask == funcTask)
+            if (funcTask.IsCompleted)
             {
                 return await funcTask.ConfigureAwait(false);
             }
-            else if (completedTask == delayTask)
+            else if (delayTask.IsCompleted)
             {
                 throw new SchedulerTimeoutException("Real time time-out; deadlock in user code?");
             }
-            else if (completedTask == simulatedTimeTask)
+            else if (simulatedTimeTask.IsCompleted)
             {
                 throw new SchedulerTimeoutException("Simulated time time-out; busy loop in user code?");
             }
             else
             {
-                throw new InvalidOperationException($"Unexpected return value from Task.WhenAny - none of the expected tasks");
+                throw new InvalidOperationException($"Unexpected return from Task.WhenAny - none of the expected tasks completed.");
             }
         }
 
-        private Task StartLoopAsync(DateTime simulatedTimeout)
+        /// <summary>
+        /// Runs the scheduler for the given number of seconds. If the task completes normally,
+        /// the clock should have advanced by the given number of seconds.
+        /// </summary>
+        /// <param name="seconds">How many seconds (in simulated time) to run for</param>
+        /// <returns>A task which will complete when scheduler has processed tasks up until
+        /// the given time.</returns>
+        public Task RunAndPauseForSeconds(int seconds) => RunAndPause(TimeSpan.FromSeconds(seconds));
```

Comment:
`RunAndPauseForSecondsAsync` ?

Reply:
I think I originally had that, then made it a void method, then made it async again but forgot to change the name back. Admittedly it's a pretty unwieldy name - if you have any other suggestions, I'd love to hear them...

Reply:
Hmm. The parameter name isn't as clear when reading code using it - especially as often unspecified time units are in milliseconds. Maybe it would be okay though...

Reply:
Yes, that's what I want to convey - it's "run for this duration" rather than "run until a delegate completes". I'm very, very much up for an entirely different name if that would be useful. (I do want some method with "seconds" though, as having to write TimeSpan.FromSeconds everywhere was really annoying.)

Let's keep thinking about names.

Reply:
What about Advance?

- `AdvanceAsync(Timespan)`
- `AdvanceSecondsAsync(int)`
- `AdvanceUntilAsync(Task)`

That said, I don't think it's bad as it is, except maybe for the missing Aysnc.

Reply:
- Advance feels like it might be skipping over rather than running along the period
- I've been avoiding Until before that suggests it's to a time rather than for a duration (although we could definitely add a method that did that)

Heck, maybe we could just overload RunAsync? Or do you think they're sufficiently different to merit different names?

Reply:
Overloading `RunAsync` sounds good to me.

Reply:
I think you haven't pushed.

Reply:
Whoops. Probably forgot to force push...

Reply:
Done now :)